Checkout: Add "Address for VAT" field to VatForm

sirbrillig opened the pull request:
#### Proposed Changes

This adds a new field to the `VatForm` component in checkout for the address (sometimes) required by our VAT processors. This field matches the one that already exists on the `/me/purchases/vat-details` page. It was left out of the checkout form intentionally in https://github.com/Automattic/wp-calypso/pull/71285 because, to quote myself,

> First, the checkout version of the form does not include the "Address" field. I did this intentionally because I felt that it cluttered the form, added friction, and may not actually be required. I also did it because once we add this form to the domain form, which already includes address fields, it's likely to feel redundant and awkward (we can't use those address fields because they are separated into address1/address2/city/state/etc.).

However, it's recently been decided that the address field is necessary in checkout as well.

<img width="570" alt="Screenshot 2023-01-27 at 2 07 25 PM" src="https://user-images.githubusercontent.com/2036909/215173979-3fca52b1-b289-4da6-a7a3-293cb53be648.png">

Part of implementing https://github.com/Automattic/payments-shilling/issues/1278

D99532-code is required for this to update the data on the server.

Depends on https://github.com/Automattic/wp-calypso/pull/72740

#### Testing Instructions

- Apply D99532-code if it is not already merged.
- On this branch, add a product to your cart and visit checkout.
- If the checkout contact/billing details step is automatically completed, click the "Edit" button to make it active.
- Select a VAT country like "United Kingdom" in the country field and a valid postal code number like `NW1 4NP`.
- Click the VAT checkbox and verify that the VAT fields appear. Set the company name to `Test company`. Set the VAT Number to a valid one like `553557881` and fill in anything in the "Address for VAT" field.
- Click continue to complete the step. (Note: this is a test number that will only work when the API is sandboxed).
- Verify that the validation is successful and that the step completes.
- Click to edit the step again.
- Verify that you can no longer edit the VAT Number but that you can edit the organization and the address.
- Visit `/me/purchases/vat-details` and verify that the form there has the same information entered in checkout.
- Return to checkout (or reload the page), edit the step again, and verify that the form is pre-populated with the VAT details you entered before.

##### Verifying the data is sent correctly

- Open your browser's devtools panel to monitor network traffic and look for transactions to the `/me/shopping-cart` endpoint.
- Click to complete the step and verify that the shopping-cart endpoint request includes the VAT `address` information in its `tax` property.

sirbrillig committed:
Add "Address for VAT" field to VatForm

sirbrillig committed:
Make VAT address field full width

sirbrillig committed:
Add address field to ResponseCartTaxData

sirbrillig committed:
Add address to shopping cart conversion functions

sirbrillig committed:
Add address to shopping-cart-endpoint tests

sirbrillig committed:
Include VAT address in transaction and cart endpoint requests

sirbrillig committed:
Add tests for VAT address

sirbrillig committed:
Fix bugs in pre-filled VAT form tests

sirbrillig commented:
This currently includes bug fixes which I've made into their own PR in https://github.com/Automattic/wp-calypso/pull/72740. I'll rebase those out when that merges.

sirbrillig committed:
Add address tests broken by rebase

pottedmeat commented:
- I tested it with a business plan upgrade
- I entered the VAT information as shown
- It saved correctly and I verified it was passed in the shopping card endpoint
- The page reloaded with the saved VAT details
- The information submitted showed up correctly in my user's VAT details page
- I was unable to edit the VAT ID
- I was able to edit the name/address and it was immediately reflected in my users's VAT details page
- When I edited the billing information back to the United States and hit Continue, it showed up with GB instead of US
- <img width="561" alt="image" src="https://user-images.githubusercontent.com/38750/215277882-1a40a6b0-39c2-432d-aefc-1f51e3757799.png">
- When I reloaded the page, the billing details showed the correct US information
- I was able to go back once again from United States to United Kingdom with no unexpected behavior

pottedmeat commented on client/my-sites/checkout/composite-checkout/components/vat-form/index.tsx:
```diff
@@ -225,6 +225,22 @@ export function VatForm( {
+			</div>
 			{ savedVatDetails.id && (
 				<div>
 					<FormSettingExplanation>
```
I actually didn't notice this when testing. Should it be above the Address field?

sirbrillig replied:
🤔 That's an interesting question. I left it below the form because it felt better to me there but we could certainly move it. I'll wait for now because I know that @michaeldcain wants to do some styling of this form and it could be worked in there.

pottedmeat reviewed (dismissed):
Code and tests look good.

sirbrillig commented:
> When I edited the billing information back to the United States and hit Continue, it showed up with GB instead of US

😬 That doesn't sound good. Thanks for spotting that! There's still something wrong with the two country fields.

sirbrillig commented:
Thanks! I refactored the way that `VatForm` pre-fills its fields to fix the issue you described in https://github.com/Automattic/wp-calypso/pull/72734#issuecomment-1407433999 and also added a test for that bug. I tried to backport those changes to https://github.com/Automattic/wp-calypso/pull/72740 but the tests are failing there and I don't have time to figure out why right now so I've just rebased this back onto trunk for now so that it can be tested.

sirbrillig committed:
Add missing vat address fields to some tests

pottedmeat commented on client/my-sites/checkout/composite-checkout/components/vat-form/index.tsx:
```diff
@@ -54,48 +54,95 @@ export function VatForm( {
 		}
-		if ( didPreFill.current ) {
+
+		if ( ! isVatSupportedFor( previousCountryCode.current ) && isVatSupportedFor( countryCode ) ) {
```
🥇

pottedmeat reviewed (approved):
Code changes look good.

Repeated my testing and it worked as expected.

sirbrillig committed:
Remove pause for validation in checkout contact step test

michaeldcain commented:
> I know that @michaeldcain wants to do some styling of this form

Don't let me be a blocker on styling. I can follow-up in another PR.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7783684

Thank you @sirbrillig for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7783684).
